ckan/lib/helpers.py

```diff
@@ -324,10 +325,14 @@ def _local_url(url_to_amend, **kw):
     url = '%s%s%s' % (root, root_path, url_path)
 
     # stop the root being added twice in redirects
-    if no_root and url_to_amend.startswith(root):
-        url = url_to_amend[len(root):]
+    if no_root and url_to_amend.startswith(root + root_path):
+        url = url[len(root + root_path):]
+
         if not default_locale:
-            url = '/%s%s' % (locale, url)
+            url = '%s%s%s' % (
+                root + root_path,
+                '' if default_locale else '/' + locale,
+                url[len(root+root_path):])
```

Comment:
this isn't right, see the root_path logic above https://github.com/ckan/ckan/pull/3035/commits/38495ab182c22b19cb8cadd767da16324be91b1c#diff-c7d65453be2c616ad1856768f048a500L310 the language is part of root path, not appended to it